File: examples/kotlin/pubsub/src/main/kotlin/ftl/pubsub/Kitchen.kt
package ftl.pubsub

import io.quarkus.logging.Log
import xyz.block.ftl.*
import java.time.OffsetDateTime

data class TimeResponse(val time: OffsetDateTime)

@Subscription(topic = NewOrderTopic::class, from = FromOffset.BEGINNING)
fun cookPizza(pizza: Pizza, topic: PizzaReadyTopic) {
  Log.infof("cooking pizza %s for %s", pizza.type, pizza.customer)
  topic.publish(pizza)
}

data class Pizza(val id : Long, val type: String, val customer : String)

@Topic
interface PizzaReadyTopic : WriteableTopic<Pizza, SinglePartitionMapper>

@Topic
interface NewOrderTopic : WriteableTopic<Pizza, SinglePartitionMapper>

File: examples/go/pubsub/kitchen.go
package pubsub

import (
	"context"

	"github.com/alecthomas/errors" // Import the FTL SDK.

	"github.com/block/ftl/go-runtime/ftl"
)

type PizzaPartitionMapper struct{}

func (PizzaPartitionMapper) PartitionKey(pizza Pizza) string {
	return pizza.Customer
}

//ftl:export
type NewOrderTopic = ftl.TopicHandle[Pizza, PizzaPartitionMapper]

//ftl:export
type PizzaReadyTopic = ftl.TopicHandle[Pizza, ftl.SinglePartitionMap[Pizza]]

type Pizza struct {
	ID       int
	Type     string
	Customer string
}

//ftl:verb
//ftl:subscribe newOrderTopic from=beginning
func CookPizza(ctx context.Context, pizza Pizza, topic PizzaReadyTopic) error {
	ftl.LoggerFromContext(ctx).Infof("Cooking pizza: %v", pizza)
	return errors.WithStack(topic.Publish(ctx, pizza))
}
